### ISSUE TITLE
Verification of JWT Token by service itself (possible with present symmetric encryption) & Consideration of asymmetric encryption for a JWT token

### ISSUE THREAD

**Issue by arvindh123** [Mainflux/mainflux-lite#1672]
**FEATURE REQUEST**

1. Is there an open issue addressing this request? If it does, please add a "+1" reaction to the
   existing issue, otherwise proceed to step 2.

2. Describe the feature you are requesting, as well as the possible use case(s) for it.

### Objective  
Verification of JWT token by service, without depending on auth service. 
Auth service will issue only token, Service by them self will verify the token 

### Present scenario 
Every request received by service need to verify token, So the service call the auth service to validate the token, 
So, Auth service will have more network traffic and computation all the time. 

### Proposal 
If service can be able to verify the JWT token signature by itself, then no need to send request to auth for verification. 

At present auth service use HMAC encryption  
To verify the JWT token by service itself, the HMAC secret key need to share with all the service.  
Distribution of HMAC secret key to service is not good way I think, since with this HMAC symmetric key we can able to verify the certificate and we can able to create JWT Token with any other tools or codes , 

Anyone who knows the HMAC symmetrical secret key can tamper the system.

## Consideration of asymmetric encryption for a JWT token (For the case service verifies JWT token by itself)
If we use asymmetric encryption key RSA256,  then we need to share only PUBLIC key with all service. 

### Implementation rough idea: 
Auth service will generate private and public key 
Private key is used by auth service to generate JWT with signature 
Private key is not shared or not distributed, It will remains in auth service

Public Key is shared to service as JSON Web Key Set - JWKS Key through HTTP endpoint 
example: http://mainflux-auth:9999/.well-known/jwks.json 
 
And also we should have two token for this case , Access token with 10 minutes or less than 10 minutes expiry time and Refresh token with 12 hours or more than it .

Access Token is to access the resource 
Refresh Token to get access token 

### Additional security by rotation of asymmetric key
In additional, in auth service, we can rotate the programmatically public and private key on daily basis, and keep the public and private key in memory so no one (even us) could tamper them,
Or we can keep it them as file whenever it is generated.

3. Indicate the importance of this feature to you (must-have, should-have, nice-to-have).
should-have

**Comment by arvindh123** [comment]
Request for Comments 
@mainflux/contributors 
@mainflux/maintainers

**Comment by drasko** [comment]
@arvindh123 in the ideal state JWTs are stateless, have short expiry date and then can be verified even on the proxy - which saves traffic and decreases latency.

However, in Mainflux we introduce also a long-live JWTs, and these can be revoked. Revocation makes JWT stateful, and their validity must now be checked in the DB (by Auth service).

This somehow beats the purpose of JWT, but if we need revocation - this is the only solution.

Maybe we can add two types of JWT checks - for those that are short-lived and those that are long-lived, and apply different strategies.

**Comment by arvindh123** [comment]
@drasko , 
I agree with your point to have two tokens with short life and long life, it also mentioned in above issue. 

Access JWT  Token  - Access token is short life , which can be stateless, 
**This can be used by Service to do self-validate without reaching Auth service.**
   
Refresh JWT Token -  Refresh Token with long life , which can be stateful , This can be use to get access token. 

To get Access JWT Token, 
The user should sent request to Auth service with Refresh Token, 
Auth service can verify the Refresh JWT token , it can also check for revocation of token and then issue Access Token to requestor.

**Comment by arvindh123** [comment]
Blocked: Waiting for new auth to merge with main branch

**Comment by arvindh123** [comment]
There is need for Stateful JWT as explained by @drasko in this ticket https://github.com/absmach/supermq/issues/2081

**Comment by felixgateru** [comment]
# Background

The issue is related to [SuperMQ Issue #2081](https://github.com/absmach/supermq/issues/2081)
The issue comments highlight the two main methods of authentication, namely **session-based** and **token-based**.

The proposed solution is dependent on which authentication methodology we choose.
In summary:

- JWT based authentication is inherently **stateless**. The advantage being that it is suitable for scalability. We can make it possible for individual services to perform authentication reducing gRPC calls to the central Auth Service. We however lose the ability to immediately revoke these tokens(especially the short lived access tokens).
- Session based authentication is inherently **stateful**. The advantage being immediate session management including logging users out and changing session details like session expiry. A possible challenge is loosing per service authentication as all requests result in calls to the Auth service.

## Proposed Implementation 1: Stateful JWT

### Overview

We will implement **stateful JWT authentication** consisting of:

- **Access Token**: Short-lived, stateless JWT.
- **Refresh Token**: Long-lived, stateful token stored in the database and revocable. It will be used to generate new access tokens.

The following link highlights the proposed architecture: [Securing our Microservices by Authentication and Authorization with JWT, Refresh Tokens and RBAC](https://medium.com/%40jose-luis-navarro/securing-our-microservices-through-authentication-and-authorization-with-jwt-refresh-tokens-and-f4439810fce7)

### Refresh Token Storage and Revocation

- Storage and revocation of refresh tokens were initially introduced in [[PR #2240](https://github.com/absmach/supermq/pull/2240)](https://github.com/absmach/supermq/pull/2240).
- The PR includes:

  - **Persistent storage** of refresh tokens.
  - **Revocation** support.
  - **Caching** to improve performance.

- The intended approach will build upon this design.

### Authentication Service

- Authentication will be handled by the **auth service** and **per service**.
- The service will use **asymmetric encryption**:

  - **Private key**: Accessible **only** by the auth service.
  - **Public key**: Distributed as a JWKS to other services through a **gRPC or HTTP endpoint** .

This follows the [[Auth0 JWKS (JSON Web Key Sets) standard](https://auth0.com/docs/secure/tokens/json-web-tokens/json-web-key-sets)](https://auth0.com/docs/secure/tokens/json-web-tokens/json-web-key-sets).

### JWKS and Caching

- The public keys will be exposed in JWKS format.
- Other services will cache the JWKS to minimize latency and reduce dependency on frequent network calls.

### Previous Work

A previous attempt for this solution is outlined in [[PR #2437](https://github.com/absmach/supermq/pull/2437/files)](https://github.com/absmach/supermq/pull/2437/files).

## Proposed Implementation 2: Session based authentication with API Gateway/BFF

To gain the advantages of session based authentication the following proposed solution has been investigated.

### Overview

All requests are funneled through an API gateway. The gateway is used to associate a client session with access tokens and refresh tokens.
An example of this is highlighted here:[The Backend for Frontend Pattern](https://auth0.com/blog/the-backend-for-frontend-pattern-bff/)

### Sessions

Sessions will be stored on high performance Redis Cache. The API Gateway returns a session cookie on login and the client uses the cookie on subsequent requests. The API gateway retrieves the associated access token, and uses the token to make a call to the API.
Sessions can be revoked by the API gateway.

### Hybrid nature

- This implementation proposes a **hybrid** scenario where JWT tokens are maintained but **sessions** are used for clients where cookies are native.

A hybrid approach providing both session and token based authentication has been noted in other projects like: 
[kratos](https://github.com/ory/kratos)

## References

I found the following links helpful in investigating the issue:

1. https://dev.to/kripanshu/stop-overthinking-authentication-a-3-question-guide-to-choosing-jwt-vs-sessions-c44
2. https://auth0.com/blog/the-backend-for-frontend-pattern-bff/
3. https://medium.com/%40jose-luis-navarro/securing-our-microservices-through-authentication-and-authorization-with-jwt-refresh-tokens-and-f4439810fce7
4. https://microservices.io/post/architecture/2025/05/28/microservices-authn-authz-part-2-authentication.html
5. https://medium.com/codetutorials/microservice-session-management-from-stateful-pain-to-stateless-gain-with-redis-and-jwt-72593a746b04
6. https://dev.to/codeparrot/jwt-vs-session-authentication-1mol
7. https://curity.io/resources/learn/the-token-handler-pattern/
